### ISSUE TITLE
Saving user-editable state in ScreenModels (or ViewModels)

### ISSUE THREAD

**Issue by bvitaliyg** [adrielcafe/voyager#323]
Hey, and thank you for this library.

There's something I noticed which might either be a documentation missing or an architectural issue.

The library allows the state saved/restored by providing saveable parameters to the `Screen` class. Later on, these parameters might be passed to `ScreenModel` to initialise it.

If the configuration changes, the `ScreenModel` survives that change and keeps the state.
If the process is killed by the system, the stack, the screen and its parameters will be serialised, saved and restored – and the new ScreenModel with the same parameters passed from the `Screen` will be initialised when the app reopens.

While this approach works perfectly in cases like that, it doesn't work when the user can edit the `ScreenModel`'s state. This state might include anything that's mutable by the user's actions, including Compose state holders, Flows or simple variables. 

When the user-editable state lives in `ScreenModel`, `ScreenModel` still survives the configuration change. When the process is killed, however, the only data that are being saved/restored are the data from `Screen`, not `ScreenModel`. Therefore, the custom state is lost.

Here's an example of a counter that does survive the configuration change, but not the process death:

```kotlin
// minimising the app and calling 'adb shell am kill appId' will simulate the process's death
class MyScreenModel(
    initialState: Int = 0,
): StateScreenModel<Int>(initialState) {

    fun increment() =
        mutableState.update { it + 1 }

    fun decrement() =
        mutableState.update { it - 1 }

}

data class MyScreen(
    val initialState: Int = 0,
) : Screen {

    @Composable
    override fun Content() {
        val screenModel = rememberScreenModel { MyScreenModel(initialState) }

        Row(Modifier.fillMaxSize()) {
            Button(onClick = screenModel::decrement) {
                Text("Decrement")
            }

            Text(text = screenModel.state.collectAsState().value.toString())

            Button(onClick = screenModel::increment) {
                Text("Increment")
            }
        }
    }

}
```

As a workaround, you can consider creating a serialisable state holder (a part of getter and setter) which will live in ScreenModel. E.g. doing something like:

```kotlin
data class MyStateHolder<T>(var state: T) : Serializable

data class MyScreen(val initialState: Int) : Screen {

    private val holder = MyStateHolder(initialState)

    @Composable
    override fun Content() {
        val screenModel = rememberScreenModel { MyScreenModel(holder) }
        // ...
    }
}
```

This approach, however, will not work well with the configuration change. As `ScreenModel` is retained, it will refer now to an object that no longer lives in active `Screen` – and therefore won't be serialised.

So far, it looks like there's something I'm missing from the documentation (and this might be something to add there), or an API design problem.

If the problem is in API, it might be resolved by providing something like this:

```kotlin
@Composable
    fun <State, ScreenModel> rememberScreenModel(
        initialState: State,
        saveableStateProvider: (ScreenModel) -> State,
        saver: Saver<State, *>,
        init: (State) -> ScreenModel,
    ) : ScreenModel
```

This method might behave following:
1. If there's an existing `ScreenModel`, return it.
2. If there's a saved state left from process death, call `init` passing the saved state.
3. If there's no saved state and the ScreenModel is created initially, call `init` passing `initialState`.
4. If the process is being terminated, extract the state from `ScreenModel` using `saveableStateProvider` and save/restore it using `Saver` (optionally).

What do you think is the correct approach? Am I missing something? Thank you.

**Comment by DevSrSouza** [comment]
ViewModel and ScreenModel are very similar on this approach, both requires you to pass a SavedStateHandle.

I really like your suggestion and have being thinking on something like that for couple months, will try to build this API when I have little free time.

**Comment by babramovitch** [comment]
@DevSrSouza I'm looking for the behaviour described above as well with user editable content in the ScreenModel not restoring.

I'm a little confused by what you mean when you say "ViewModel and ScreenModel are very similar on this approach, both requires you to pass a SavedStateHandle" as I can't find any reference to being able to use a SavedStateHandle with ScreenModel anywhere?

Is that just a comment that it's required but it's not implemented yet?

**Comment by bvitaliyg** [comment]
@DevSrSouza yes, the exact mechanism to support state keeping is the details. What I suggested is just an option I'd consider idiomatic for Compose, but in reality, any mechanism would work, as long as it's there. `SavedStateHandle` is also a great option, and it's idiomatic for `ViewModels` in general.
